### PR TITLE
Add a lot more user friendly launch plan creating function

### PR DESCRIPTION
**Pull request by wild-endeavor**
Signed-off-by: wild-endeavor <wild-endeavor@users.noreply.github.com>

# TL;DR
It was decided that the existing functions for creating launch plans were not user-friendly.  The `get_default_launch_plan` function required a context, but it's supposed to be the `FlyteContext` not the user-facing context (`ExecutionParameters`). The regular `create` function requires a name all the time. This PR introduces a new function that is
* cached by default
* Doesn't need the name. If no name is provided, it's assumed the default launch plan is what you're after.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
See code.

## Follow-up issue
https://github.com/flyteorg/flyte/issues/892

### CODE REVIEW

File: flytekit/core/launch_plan.py [hunk 1]
```diff
@@ -20,6 +20,15 @@ class LaunchPlan(object):
 
     @staticmethod
```
Comment (katrogan) [inline flytekit/core/launch_plan.py:21]: want to mark this deprecated?

**Review by katrogan — DISMISSED** [review]

File: flytekit/core/launch_plan.py (hunk 1 above)
Comment (katrogan) [reply flytekit/core/launch_plan.py:21]: oh nvm, still being used below

**Review by cosmicBboy — APPROVED** [review]

**Comment by kumare3** [comment]
haha - slightly more?

File: flytekit/core/launch_plan.py [hunk 2]
```diff
@@ -93,6 +102,61 @@ def create(
...
         return lp
 
+    @classmethod
+    def get_or_create(
+        cls,
+        workflow: _annotated_workflow.WorkflowBase,
+        name: Optional[str] = None,
+        default_inputs: Dict[str, Any] = None,
+        fixed_inputs: Dict[str, Any] = None,
+        schedule: _schedule_model.Schedule = None,
```
Comment (kumare3) [inline flytekit/core/launch_plan.py:112]: should we put all of this 
`schedule, notifications, auth_role` etc in a dataclass? something like meta etc
as we keep adding things here?

File: flytekit/core/launch_plan.py [hunk 3]
```diff
@@ -93,6 +102,61 @@ def create(
...
+        """
+        if name is None and (
+            default_inputs is not None
+            or fixed_inputs is not None
+            or schedule is not None
+            or notifications is not None
+            or auth_role is not None
+        ):
+            raise ValueError(
+                "When get_or_create'ing the default launch plan, no arguments other than the workflow may be specified"
```
Comment (kumare3) [inline flytekit/core/launch_plan.py:142]:
```suggestion
                "Only named launchplans can be created that have other properties. Drop the name if you want to create a default launchplan. Default launchplans cannot have any other associations"
```

Reword better, i added some stuff

File: flytekit/core/launch_plan.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/core/launch_plan.py:112]: Maybe, but then the user has to make one more object to make this.  Can we hold off until we have one more thing?  It makes sense I guess if they're creating fifteen launch plans, and set all three settings with the same value - then they can create the meta once and use fifteen times, but i feel like that's not a common use case.

Comment (kumare3) [reply flytekit/core/launch_plan.py:112]: if you do later, then it becomes a breaking change. I guess you are right though, this list since its optional can keep on increasing. 

Just that your first check for nil, has to keep evolving as well

**Review by kumare3 — APPROVED** [review]
